pkg/semtech-loramac/contrib/semtech_loramac_getset.c

```diff
@@ -27,217 +29,305 @@
+    mutex_unlock(&mac->lock);
+    return enable;
 }
 
-void semtech_loramac_set_netid(uint32_t netid)
+void semtech_loramac_set_netid(semtech_loramac_t *mac, uint32_t netid)
 {
+    mutex_lock(&mac->lock);
     DEBUG("[semtech-loramac] set NetID %lu\n", netid);
     MibRequestConfirm_t mibReq;
     mibReq.Type = MIB_NET_ID;
     mibReq.Param.NetID = netid;
     LoRaMacMibSetRequestConfirm(&mibReq);
+    mutex_unlock(&mac->lock);
 }
 
-uint32_t semtech_loramac_get_netid(void)
+uint32_t semtech_loramac_get_netid(semtech_loramac_t *mac)
 {
+    mutex_lock(&mac->lock);
+    uint32_t netid;
     DEBUG("[semtech-loramac] get NetID\n");
     MibRequestConfirm_t mibReq;
     mibReq.Type = MIB_NET_ID;
     LoRaMacMibGetRequestConfirm(&mibReq);
-    return mibReq.Param.NetID;
+    netid = mibReq.Param.NetID;
+    mutex_unlock(&mac->lock);
+    return netid;
 }
 
-void semtech_loramac_set_tx_power(uint8_t power)
+void semtech_loramac_set_tx_power(semtech_loramac_t *mac, uint8_t power)
 {
+    mutex_lock(&mac->lock);
     DEBUG("[semtech-loramac] set TX power %d\n", power);
     MibRequestConfirm_t mibReq;
     mibReq.Type = MIB_CHANNELS_TX_POWER;
     mibReq.Param.ChannelsTxPower = power;
     LoRaMacMibSetRequestConfirm(&mibReq);
+    mutex_unlock(&mac->lock);
 }
 
-uint8_t semtech_loramac_get_tx_power(void)
+uint8_t semtech_loramac_get_tx_power(semtech_loramac_t *mac)
 {
+    mutex_lock(&mac->lock);
+    uint8_t tx_power;
     DEBUG("[semtech-loramac] get TX power\n");
     MibRequestConfirm_t mibReq;
     mibReq.Type = MIB_CHANNELS_TX_POWER;
     LoRaMacMibGetRequestConfirm(&mibReq);
-    return (uint8_t)mibReq.Param.ChannelsTxPower;
+    tx_power = (uint8_t)mibReq.Param.ChannelsTxPower;
+    mutex_unlock(&mac->lock);
+    return tx_power;
+}
+
+void semtech_loramac_set_tx_port(semtech_loramac_t *mac, uint8_t port)
+{
+    mac->port = port;
+}
+
+uint8_t semtech_loramac_get_tx_port(semtech_loramac_t *mac)
+{
+    return mac->port;
+}
+
+void semtech_loramac_set_tx_mode(semtech_loramac_t *mac, uint8_t mode)
+{
+    mac->cnf = mode;
 }
 
-void semtech_loramac_set_rx2_freq(uint8_t freq)
+uint8_t semtech_loramac_get_tx_mode(semtech_loramac_t *mac)
 {
-    Rx2ChannelParams_t params;
+    return mac->cnf;
+}
+
+void semtech_loramac_set_rx2_freq(semtech_loramac_t *mac, uint32_t freq)
+{
+    mutex_lock(&mac->lock);
+    DEBUG("[semtech-loramac] setting RX2 freq to %lu\n", freq);
+    Rx2ChannelParams_t p;
     MibRequestConfirm_t mibReq;
     mibReq.Type = MIB_RX2_DEFAULT_CHANNEL;
     LoRaMacMibGetRequestConfirm(&mibReq);
-    params.Frequency = freq;
-    params.Datarate = mibReq.Param.Rx2DefaultChannel.Datarate;
-    _loramac_set_rx2_params(params.Frequency, params.Datarate);
+    p.Frequency = freq;
+    p.Datarate = mibReq.Param.Rx2DefaultChannel.Datarate;
+    semtech_loramac_channel_params_t params;
+    params.frequency = freq;
+    params.datarate = p.Datarate;
+    semtech_loramac_set_rx2_params(mac, params);
```

Comment:
Just wondering, isn't a problem to call this function which will re-call `mutex_lock(&mac->lock)` before we unlock it ?

Reply:
Yes it is. Good finding. I admit I didn't test this function. If you try `semtech_loramac_set_rx2_freq` or `semtech_loramac_set_rx2_dr` from the shell application, I think the command will just hang.